src/tokens/ERC721.sol

```diff
@@ -0,0 +1,312 @@
+        keccak256("Permit(address owner,address spender,uint256 nonce,uint256 deadline)");
+    
+    uint256 internal immutable INITIAL_CHAIN_ID;
+
+    bytes32 internal immutable INITIAL_DOMAIN_SEPARATOR;
+
+    mapping(uint256 => uint256) public nonces;
+
+    mapping(address => uint256) public noncesForAll;
+    
+    /*///////////////////////////////////////////////////////////////
+                            CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+    
+    constructor(string memory _name, string memory _symbol) {
+        name = _name;
+        symbol = _symbol;
+        
+        INITIAL_CHAIN_ID = block.chainid;
+        INITIAL_DOMAIN_SEPARATOR = computeDomainSeparator();
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            ERC-20-LIKE LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function transfer(address to, uint256 tokenId) public virtual returns (bool success) {
+        require(msg.sender == ownerOf[tokenId], "NOT_OWNER");
+        
+        // Cannot overflow because because ownership is checked
+        // against decrement, and sum of all user
+        // balances can't exceed the max uint256 value.
+        unchecked {
+            balanceOf[msg.sender]--; 
+        
+            balanceOf[to]++;
+        }
+        
+        delete getApproved[tokenId];
+        
+        ownerOf[tokenId] = to;
+        
+        emit Transfer(msg.sender, to, tokenId); 
+        
+        success = true;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            ERC-721 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function supportsInterface(bytes4 interfaceId) public pure virtual returns (bool supported) {
+        supported = interfaceId == 0x80ac58cd || interfaceId == 0x5b5e139f || interfaceId == 0x01ffc9a7;
+    }
+    
+    function approve(address spender, uint256 tokenId) public virtual {
+        address owner = ownerOf[tokenId];
+        
+        require(msg.sender == owner || isApprovedForAll[owner][msg.sender], "NOT_APPROVED");
+        
+        getApproved[tokenId] = spender;
+        
+        emit Approval(owner, spender, tokenId); 
+    }
+    
+    function setApprovalForAll(address operator, bool approved) public virtual {
+        isApprovedForAll[msg.sender][operator] = approved;
+        
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+    
+    function transferFrom(
+        address, 
+        address to, 
+        uint256 tokenId
+    ) public virtual {
+        address owner = ownerOf[tokenId];
+        
+        require(
+            msg.sender == owner 
+            || msg.sender == getApproved[tokenId]
+            || isApprovedForAll[owner][msg.sender], 
+            'NOT_APPROVED'
+        );
+        
+        // Cannot under/overflow because ownership is checked
```

Comment:
Hey! We are using a slightly modified version of this on EtherOrcs and we just uncovered a vulnerability with this pattern. 

I two parties approve the same contract, one can transfer the token of the other because `from` is not checked here.